File: internal/snapshot/compare_test.go
package snapshot

import (
	"reflect"
	"testing"
)

func sec(opts func(*Section)) Section {
	s := Section{}
	if opts != nil {
		opts(&s)
	}
	return s
}
func item(raw string, cols ...string) Item {
	if len(cols) == 0 {
		cols = []string{raw}
	}
	return Item{Raw: raw, Columns: cols}
}

func TestCompareOrientation(t *testing.T) {
	d := Compare(
		Snapshot{"onlyLeft": {}},
		Snapshot{"onlyRight": {}},
	)
	if d["onlyLeft"].Status != StatusAdded {
		t.Errorf("left-only should be added, got %s", d["onlyLeft"].Status)
	}
	if d["onlyRight"].Status != StatusRemoved {
		t.Errorf("right-only should be removed, got %s", d["onlyRight"].Status)
	}
}

func TestCompareItems(t *testing.T) {
	left := Snapshot{"s": {Items: []Item{item("keep"), item("ladd")}}}
	right := Snapshot{"s": {Items: []Item{item("keep"), item("radd")}}}
	d := Compare(left, right)["s"]
	if d.Status != StatusChanged {
		t.Errorf("want changed, got %s", d.Status)
	}
	if !reflect.DeepEqual(d.Items.Added, []string{"ladd"}) {
		t.Errorf("added: %v", d.Items.Added)
	}
	if !reflect.DeepEqual(d.Items.Removed, []string{"radd"}) {
		t.Errorf("removed: %v", d.Items.Removed)
	}
	if !reflect.DeepEqual(d.Items.Common, []string{"keep"}) {
		t.Errorf("common: %v", d.Items.Common)
	}
}

func TestComparePairs(t *testing.T) {
	left := Snapshot{"s": {Pairs: map[string]string{"same": "1", "chg": "L", "onlyL": "l"}}}
	right := Snapshot{"s": {Pairs: map[string]string{"same": "1", "chg": "R", "onlyR": "r"}}}
	d := Compare(left, right)["s"]
	if !reflect.DeepEqual(d.Pairs.Added, map[string]string{"onlyL": "l"}) {
		t.Errorf("added: %v", d.Pairs.Added)
	}
	if !reflect.DeepEqual(d.Pairs.Removed, map[string]string{"onlyR": "r"}) {
		t.Errorf("removed: %v", d.Pairs.Removed)
	}
	if d.Pairs.Changed["chg"] != (PairChange{Left: "L", Right: "R"}) {
		t.Errorf("changed: %v", d.Pairs.Changed)
	}
	if !reflect.DeepEqual(d.Pairs.Common, map[string]string{"same": "1"}) {
		t.Errorf("common: %v", d.Pairs.Common)
	}
}

func TestCompareCommonOnlyIsEqual(t *testing.T) {
	c := "c"
	s := Snapshot{"s": {Items: []Item{item("a")}, Pairs: map[string]string{"k": "v"}, Content: &c}}
	clone := Snapshot{"s": {Items: []Item{item("a")}, Pairs: map[string]string{"k": "v"}, Content: &c}}
	if got := Compare(s, clone)["s"].Status; got != StatusEqual {
		t.Errorf("identical sections should be equal, got %s", got)
	}
}

func TestCompareContent(t *testing.T) {
	x := "x"
	d := Compare(Snapshot{"s": {Content: &x}}, Snapshot{"s": {}})["s"]
	if !d.Content.Changed || d.Content.Left == nil || d.Content.Right != nil {
		t.Errorf("content diff wrong: %+v", d.Content)
	}
	if d.Status != StatusChanged {
		t.Errorf("want changed, got %s", d.Status)
	}
}

func TestFormatChangesOnly(t *testing.T) {
	// identical → empty output (so callers can print "No differences")
	s := Snapshot{"x": {Items: []Item{item("i")}, Pairs: map[string]string{"k": "v"}}}
	clone := Snapshot{"x": {Items: []Item{item("i")}, Pairs: map[string]string{"k": "v"}}}
	if out := Compare(s, clone).Format(FormatOptions{ChangesOnly: true}); out != "" {
		t.Errorf("identical snapshots should render empty, got %q", out)
	}
}

func TestFormatChangesOnlyKeepsChanges(t *testing.T) {
	left := Snapshot{
		"keep": {Items: []Item{item("a")}},
		"chg":  {Items: []Item{item("x"), item("only")}},
	}
	right := Snapshot{
		"keep": {Items: []Item{item("a")}},
		"chg":  {Items: []Item{item("x")}},
	}
	out := Compare(left, right).Format(FormatOptions{ChangesOnly: true})
	want := "[chg]\n  + only  (only in left)" // 'keep' (equal) gone, common 'x' gone
	if out != want {
		t.Errorf("changesOnly output:\n got %q\nwant %q", out, want)
	}
}

func TestFormatColorAndLabels(t *testing.T) {
	out := Compare(Snapshot{"x": {Items: []Item{item("i")}}}, Snapshot{}).
		Format(FormatOptions{LeftLabel: "new", RightLabel: "old", Color: false})
	want := "+ [x]  (only in new)\n  + i  (only in new)"
	if out != want {
		t.Errorf("\n got %q\nwant %q", out, want)
	}

	colored := Compare(Snapshot{"x": {Items: []Item{item("i")}}}, Snapshot{"x": {}}).
		Format(FormatOptions{Color: true})
	if !contains(colored, "\x1b[32m+ i\x1b[0m") {
		t.Errorf("expected green add, got %q", colored)
	}
}

func contains(s, sub string) bool {
	return len(s) >= len(sub) && (func() bool {
		for i := 0; i+len(sub) <= len(s); i++ {
			if s[i:i+len(sub)] == sub {
				return true
			}
		}
		return false
	}())
}

File: tests/snapshot/compare.test.ts
import { test, expect, describe } from "bun:test";
import { compareSnapshots, formatDiff } from "../../src/snapshot";
import type { Snapshot } from "../../src/snapshot";

const sec = (name: string, extra: Partial<Snapshot[string]> = {}): Snapshot[string] => ({
  name,
  pairs: {},
  items: [],
  content: null,
  ...extra,
});
const item = (raw: string, ...columns: string[]) => ({ raw, columns: columns.length ? columns : [raw] });

describe("compareSnapshots (orientation: first arg = left = added)", () => {
  test("left-only section is 'added', right-only is 'removed'", () => {
    const diff = compareSnapshots({ onlyLeft: sec("onlyLeft") }, { onlyRight: sec("onlyRight") });
    expect(diff.sections.onlyLeft.status).toBe("added");
    expect(diff.sections.onlyRight.status).toBe("removed");
  });

  test("items diff keys on raw: added=left-only, removed=right-only, common=both", () => {
    const left = { s: sec("s", { items: [item("keep"), item("ladd")] }) };
    const right = { s: sec("s", { items: [item("keep"), item("radd")] }) };
    const d = compareSnapshots(left, right).sections.s;
    expect(d.status).toBe("changed");
    expect(d.items).toEqual({ added: ["ladd"], removed: ["radd"], common: ["keep"] });
  });

  test("pairs diff: added(left-only), removed(right-only), changed(both differ), common(both equal)", () => {
    const left = { s: sec("s", { pairs: { same: "1", chg: "L", onlyL: "l" } }) };
    const right = { s: sec("s", { pairs: { same: "1", chg: "R", onlyR: "r" } }) };
    const d = compareSnapshots(left, right).sections.s;
    expect(d.pairs.added).toEqual({ onlyL: "l" });
    expect(d.pairs.removed).toEqual({ onlyR: "r" });
    expect(d.pairs.changed).toEqual({ chg: { left: "L", right: "R" } });
    expect(d.pairs.common).toEqual({ same: "1" });
  });

  test("items in common alone do NOT mark a section changed", () => {
    const same = { s: sec("s", { items: [item("a")], pairs: { k: "v" }, content: "c" }) };
    expect(compareSnapshots(same, structuredClone(same)).sections.s.status).toBe("equal");
  });

  test("content diff is strict string|null inequality", () => {
    const d = compareSnapshots({ s: sec("s", { content: "x" }) }, { s: sec("s", { content: null }) }).sections.s;
    expect(d.content).toEqual({ left: "x", right: null, changed: true });
    expect(d.status).toBe("changed");
  });

  test("section name union is left-first", () => {
    const diff = compareSnapshots({ b: sec("b"), a: sec("a") }, { c: sec("c"), a: sec("a") });
    expect(Object.keys(diff.sections)).toEqual(["b", "a", "c"]);
  });
});

describe("formatDiff (verified byte-identical to @dotformat/core)", () => {
  test("color:false renders the fixed emission order with correct +/-/~/= and labels", () => {
    const left = {
      pkg: sec("pkg", { items: [item("hello"), item("world")], pairs: { a: "1", b: "L" }, content: "X" }),
    };
    const right = { pkg: sec("pkg", { items: [item("world")], pairs: { a: "1", b: "R", c: "2" }, content: "Y" }) };
    const out = formatDiff(compareSnapshots(left, right), { color: false });
    expect(out).toBe(
      [
        "[pkg]",
        "  + hello  (only in left)", // item left-only = added
        "  = world", // item common
        "  - c = 2  (only in right)", // pair right-only = removed
        "  ~ b = L → R", // pair changed (U+2192)
        "  = a = 1", // pair common
        "  ~ content changed",
      ].join("\n"),
    );
  });

  test("added/removed section headers use leftLabel/rightLabel", () => {
    const out = formatDiff(compareSnapshots({ x: sec("x", { items: [item("i")] }) }, {}), {
      leftLabel: "new",
      rightLabel: "old",
      color: false,
    });
    expect(out).toBe(["+ [x]  (only in new)", "  + i  (only in new)"].join("\n"));
  });

  test("color:true wraps in ANSI codes; no trailing newline", () => {
    const out = formatDiff(compareSnapshots({ x: sec("x", { items: [item("i")] }) }, { x: sec("x") }), { color: true });
    expect(out).toContain("\x1b[32m+ i\x1b[0m"); // green add
    expect(out.endsWith("\n")).toBe(false);
  });

  test("equal sections are still emitted (dim = lines) — no filtering", () => {
    const s = { x: sec("x", { items: [item("i")], pairs: { k: "v" } }) };
    const out = formatDiff(compareSnapshots(s, structuredClone(s)), { color: false });
    expect(out).toBe(["[x]", "  = i", "  = k = v"].join("\n"));
  });
});
